Error in @ConfigurationBuilder when updating to 5.1

Clone branch associated with PR: 

https://github.com/micronaut-projects/micronaut-email/pull/629

./gradlew compileJava

```
Caused by: java.lang.IllegalStateException: Method connectionTimeout parameters: 1 doesn't match values provided: 2
        at io.micronaut.sourcegen.model.ExpressionDef$InvokeInstanceMethod.<init>(ExpressionDef.java:1082)
        at io.micronaut.sourcegen.model.ExpressionDef.invoke(ExpressionDef.java:769)
        at io.micronaut.inject.writer.BeanDefinitionWriter.lambda$getConfigBuilderPointStatement$1(BeanDefinitionWriter.java:1611)
        at io.micronaut.sourcegen.model.ExpressionDef.newLocal(ExpressionDef.java:418)
        at io.micronaut.inject.writer.BeanDefinitionWriter.lambda$getConfigBuilderPointStatement$0(BeanDefinitionWriter.java:1588)
        at io.micronaut.sourcegen.model.ExpressionDef.newLocal(ExpressionDef.java:418)
        at io.micronaut.inject.writer.BeanDefinitionWriter.getConfigBuilderPointStatement(BeanDefinitionWriter.java:1585)
        at io.micronaut.inject.writer.BeanDefinitionWriter.getBuilderMethodStatements(BeanDefinitionWriter.java:1563)
        at io.micronaut.inject.writer.BeanDefinitionWriter.lambda$getInjectMethodConfigurationBuilder$1(BeanDefinitionWriter.java:1552)
        at io.micronaut.sourcegen.model.ExpressionDef.newLocal(ExpressionDef.java:418)
        at io.micronaut.inject.writer.BeanDefinitionWriter.getInjectMethodConfigurationBuilder(BeanDefinitionWriter.java:1551)
        at io.micronaut.inject.writer.BeanDefinitionWriter.lambda$getInjectMethod$1(BeanDefinitionWriter.java:1455)
        at io.micronaut.sourcegen.model.ExpressionDef.newLocal(ExpressionDef.java:418)
        at io.micronaut.inject.writer.BeanDefinitionWriter.lambda$getInjectMethod$0(BeanDefinitionWriter.java:1432)
        at io.micronaut.sourcegen.model.MethodDef$MethodDefBuilder.build(MethodDef.java:790)
        at io.micronaut.sourcegen.model.MethodDef$MethodDefBuilder.build(MethodDef.java:837)
        at io.micronaut.inject.writer.BeanDefinitionWriter.getInjectMethod(BeanDefinitionWriter.java:1431)
        at io.micronaut.inject.writer.BeanDefinitionWriter.build(BeanDefinitionWriter.java:1232)
        at io.micronaut.inject.writer.BeanDefinitionWriter.build(BeanDefinitionWriter.java:225)
        at io.micronaut.inject.processing.DeclaredBeanElementCreator.buildInternal(DeclaredBeanElementCreator.java:119)
        at io.micronaut.inject.processing.AbstractBeanElementCreator.build(AbstractBeanElementCreator.java:54)
        at io.micronaut.inject.processing.BeanDefinitionCreatorFactory.produceInternal(BeanDefinitionCreatorFactory.java:56)
        at io.micronaut.inject.processing.BeanDefinitionCreatorFactory.produce(BeanDefinitionCreatorFactory.java:46)
```
### Version

5.1.0

--- a/core-processor/src/main/java/io/micronaut/inject/writer/BeanDefinitionWriter.java
+++ b/core-processor/src/main/java/io/micronaut/inject/writer/BeanDefinitionWriter.java
@@ -1577,7 +1577,7 @@ public final class BeanDefinitionWriter implements BeanElement, Toggleable, Elem
         String propertyName = injectionPoint.propertyName();
         ClassElement propertyType = injectionPoint.type();
         Map<String, ClassElement> generics = propertyType.getTypeArguments();
-        boolean isDurationWithTimeUnit = propertyType.getName().equals(Duration.class.getName());
+        boolean isDurationWithTimeUnit = isDurationWithTimeUnit(methodElement, propertyType);
 
         // Optional optional = AbstractBeanDefinition.getValueForPath(...)
         String localName = builderVar.name() + "_optional" + NameUtils.capitalize(propertyName);
@@ -1613,6 +1613,14 @@ public final class BeanDefinitionWriter implements BeanElement, Toggleable, Elem
                         })
                     );
             });
+    }
+
+    private boolean isDurationWithTimeUnit(MethodElement methodElement, ClassElement propertyType) {
+        ParameterElement[] parameters = methodElement.getParameters();
+        return parameters.length == 2
+            && propertyType.getName().equals(Duration.class.getName())
+            && parameters[0].getType().getSimpleName().equals("long")
+            && parameters[1].getType().isAssignable(TimeUnit.class);
     }
 
     private StatementDef injectAndReturn(VariableDef.This aThis,

--- a/core-processor/src/main/java/io/micronaut/inject/writer/BeanDefinitionWriter.java
+++ b/core-processor/src/main/java/io/micronaut/inject/writer/BeanDefinitionWriter.java
@@ -1619,7 +1619,7 @@ public final class BeanDefinitionWriter implements BeanElement, Toggleable, Elem
         ParameterElement[] parameters = methodElement.getParameters();
         return parameters.length == 2
             && propertyType.getName().equals(Duration.class.getName())
-            && parameters[0].getType().getSimpleName().equals("long")
+            && parameters[0].getType().getSimpleName().equalsIgnoreCase("long")
             && parameters[1].getType().isAssignable(TimeUnit.class);
     }
 

--- a/inject-groovy/src/test/groovy/io/micronaut/inject/configproperties/ConfigurationPropertiesBuilderSpec.groovy
+++ b/inject-groovy/src/test/groovy/io/micronaut/inject/configproperties/ConfigurationPropertiesBuilderSpec.groovy
@@ -84,7 +84,7 @@ class TestAA {
         bean.test.bar == null
 
         cleanup:
-        applicationContext.close()
+        applicationContext?.close()
     }
 
     void "test configuration builder with includes"() {
@@ -139,7 +139,7 @@ class TestA {
         bean.test.bar == null
 
         cleanup:
-        applicationContext.close()
+        applicationContext?.close()
     }
 
     void "test configuration builder with factory method and properties"() {
@@ -179,7 +179,7 @@ class TestB {
         bean.test.bar == 'good'
 
         cleanup:
-        applicationContext.close()
+        applicationContext?.close()
     }
 
     void "test catch and log NoSuchMethodError for when underlying builder changes"() {
@@ -212,7 +212,7 @@ class TestC {
         factory.instantiate(applicationContext)
 
         cleanup:
-        applicationContext.close()
+        applicationContext?.close()
     }
 
     void "test with groovy properties"() {
@@ -261,7 +261,7 @@ class TestD {
         test.baz == null //deprecated properties are ignored
 
         cleanup:
-        applicationContext.close()
+        applicationContext?.close()
     }
 
     void "test different inject types for config properties"() {
@@ -310,7 +310,7 @@ class Neo4jProperties {
         config.logLeakedSessions()
 
         cleanup:
-        applicationContext.close()
+        applicationContext?.close()
     }
 
     void "test specifying a configuration prefix"() {
@@ -361,7 +361,7 @@ class Neo4jProperties {
         config.logLeakedSessions()
 
         cleanup:
-        applicationContext.close()
+        applicationContext?.close()
     }
 
     void "test specifying a configuration prefix with value"() {
@@ -412,7 +412,7 @@ class Neo4jProperties {
         config.logLeakedSessions()
 
         cleanup:
-        applicationContext.close()
+        applicationContext?.close()
     }
 
     void "test builder method long and TimeUnit arguments"() {
@@ -457,7 +457,7 @@ class Neo4jProperties {
         config.idleTimeBeforeConnectionTest() == 6000
 
         cleanup:
-        applicationContext.close()
+        applicationContext?.close()
     }
 
     void "test using a builder that is marked final"() {
@@ -496,7 +496,7 @@ class Neo4jProperties {
         config.idleTimeBeforeConnectionTest() == 17000
 
         cleanup:
-        applicationContext.close()
+        applicationContext?.close()
     }
 
 
@@ -571,7 +571,7 @@ class DefaultConnectionPool implements ConnectionPool {
         testPropBean.builder.build().getMaxConcurrency() == 123
 
         cleanup:
-        ctx.close()
+        ctx?.close()
     }
 
     void "test configuration builder with both getters and setters and no prefixes"() {
@@ -651,6 +651,6 @@ class DefaultConnectionPool implements ConnectionPool {
         testPropBean.builder.build().getMaxConcurrency() == 123
 
         cleanup:
-        ctx.close()
+        ctx?.close()
     }
 }

--- a/inject-groovy/src/test/groovy/io/micronaut/inject/configproperties/ConfigurationPropertiesBuilderSpec.groovy
+++ b/inject-groovy/src/test/groovy/io/micronaut/inject/configproperties/ConfigurationPropertiesBuilderSpec.groovy
@@ -82,6 +82,9 @@ class TestAA {
         then:
         bean.test.foo == 'good'
         bean.test.bar == null
+
+        cleanup:
+        applicationContext.close()
     }
 
     void "test configuration builder with includes"() {
@@ -134,8 +137,10 @@ class TestA {
         then:
         bean.test.foo == 'good'
         bean.test.bar == null
-    }
 
+        cleanup:
+        applicationContext.close()
+    }
 
     void "test configuration builder with factory method and properties"() {
         given:
@@ -172,6 +177,9 @@ class TestB {
 
         then:
         bean.test.bar == 'good'
+
+        cleanup:
+        applicationContext.close()
     }
 
     void "test catch and log NoSuchMethodError for when underlying builder changes"() {
@@ -202,8 +210,10 @@ class TestC {
                 'test.foo':'good',
         )
         factory.instantiate(applicationContext)
-    }
 
+        cleanup:
+        applicationContext.close()
+    }
 
     void "test with groovy properties"() {
         given:
@@ -249,6 +259,9 @@ class TestD {
         test.foo == 'good'
         test.bar == 10
         test.baz == null //deprecated properties are ignored
+
+        cleanup:
+        applicationContext.close()
     }
 
     void "test different inject types for config properties"() {
@@ -295,6 +308,9 @@ class Neo4jProperties {
         config.maxConnectionPoolSize() == 2
         !config.encrypted()
         config.logLeakedSessions()
+
+        cleanup:
+        applicationContext.close()
     }
 
     void "test specifying a configuration prefix"() {
@@ -343,6 +359,9 @@ class Neo4jProperties {
         config.maxConnectionPoolSize() == 2
         !config.encrypted()
         config.logLeakedSessions()
+
+        cleanup:
+        applicationContext.close()
     }
 
     void "test specifying a configuration prefix with value"() {
@@ -391,6 +410,9 @@ class Neo4jProperties {
         config.maxConnectionPoolSize() == 2
         !config.encrypted()
         config.logLeakedSessions()
+
+        cleanup:
+        applicationContext.close()
     }
 
     void "test builder method long and TimeUnit arguments"() {
@@ -433,6 +455,9 @@ class Neo4jProperties {
 
         then:
         config.idleTimeBeforeConnectionTest() == 6000
+
+        cleanup:
+        applicationContext.close()
     }
 
     void "test using a builder that is marked final"() {
@@ -469,6 +494,9 @@ class Neo4jProperties {
 
         then:
         config.idleTimeBeforeConnectionTest() == 17000
+
+        cleanup:
+        applicationContext.close()
     }
 
 
@@ -541,6 +569,9 @@ class DefaultConnectionPool implements ConnectionPool {
         then:
         noExceptionThrown()
         testPropBean.builder.build().getMaxConcurrency() == 123
+
+        cleanup:
+        ctx.close()
     }
 
     void "test configuration builder with both getters and setters and no prefixes"() {
@@ -618,5 +649,8 @@ class DefaultConnectionPool implements ConnectionPool {
         then:
         noExceptionThrown()
         testPropBean.builder.build().getMaxConcurrency() == 123
+
+        cleanup:
+        ctx.close()
     }
 }

--- a/inject-java/src/test/groovy/io/micronaut/inject/configproperties/ConfigurationPropertiesBuilderSpec.groovy
+++ b/inject-java/src/test/groovy/io/micronaut/inject/configproperties/ConfigurationPropertiesBuilderSpec.groovy
@@ -21,6 +21,8 @@ import io.micronaut.inject.BeanDefinition
 import io.micronaut.inject.InstantiatableBeanDefinition
 import org.neo4j.driver.Config
 
+import java.time.Duration
+
 /**
  * @author Graeme Rocher
  * @since 1.0
@@ -535,6 +537,65 @@ class Neo4jProperties {
 
         then:
         config.idleTimeBeforeConnectionTest() == 6000
+    }
+
+    void "test builder method Duration argument"() {
+        given:
+        BeanDefinition beanDefinition = buildBeanDefinition('test.DurationProperties', '''
+package test;
+
+import io.micronaut.context.annotation.*;
+import java.time.Duration;
+
+@ConfigurationProperties("duration.test")
+class DurationProperties {
+
+    @ConfigurationBuilder(
+        prefixes="with"
+    )
+    Options.Builder options = Options.builder();
+
+}
+
+class Options {
+    private final Duration connectionTimeout;
+
+    Options(Duration connectionTimeout) {
+        this.connectionTimeout = connectionTimeout;
+    }
+
+    public Duration getConnectionTimeout() {
+        return connectionTimeout;
+    }
+
+    static Builder builder() {
+        return new Builder();
+    }
+
+    static final class Builder {
+        private Duration connectionTimeout;
+
+        public Builder withConnectionTimeout(Duration connectionTimeout) {
+            this.connectionTimeout = connectionTimeout;
+            return this;
+        }
+
+        public Options build() {
+            return new Options(connectionTimeout);
+        }
+    }
+}
+''')
+
+        when:
+        InstantiatableBeanDefinition factory = beanDefinition
+        ApplicationContext applicationContext = ApplicationContext.run(
+                'duration.test.connection-timeout': '6s'
+        )
+        def bean = factory.instantiate(applicationContext)
+
+        then:
+        bean.options.build().getConnectionTimeout() == Duration.ofSeconds(6)
     }
 
     void "test using a builder that is marked final"() {

--- a/inject-java/src/test/groovy/io/micronaut/inject/configproperties/ConfigurationPropertiesBuilderSpec.groovy
+++ b/inject-java/src/test/groovy/io/micronaut/inject/configproperties/ConfigurationPropertiesBuilderSpec.groovy
@@ -596,6 +596,9 @@ class Options {
 
         then:
         bean.options.build().getConnectionTimeout() == Duration.ofSeconds(6)
+
+        cleanup:
+        applicationContext?.close()
     }
 
     void "test using a builder that is marked final"() {

